### ISSUE TITLE
gie test failures

### ISSUE THREAD

**Issue by sebastic** [OSGeo/PROJ#769]
Various gie tests fail for the 5.0.0-rc1 Debian package build:
```
make[5]: Entering directory '/build/proj-5.0.0~rc1/test'
../src/gie gie/*.gie
-------------------------------------------------------------------------------
Reading file 'gie/4D-API_cs2cs-style.gie'
-------------------------------------------------------------------------------
proj=latlong ellps=WGS84 towgs84=0.676780,0.654950,-0.528270,-0.022742...
-------------------------------------------------------------------------------
Invalid operation definition in line no. 26: (null)
-------------------------------------------------------------------------------
proj=latlong ellps=WGS84 towgs84=0.676780,0.654950,-0.528270,-0.022742...
-------------------------------------------------------------------------------
Invalid operation definition in line no. 26: (null)
-------------------------------------------------------------------------------
proj=latlong ellps=intl towgs84=081.07030,089.36030,115.75260,000.4848...
-------------------------------------------------------------------------------
Invalid operation definition in line no. 49: (null)
-------------------------------------------------------------------------------
proj=latlong nadgrids=nzgd2kgrid0005.gsb ellps=GRS80                  
-------------------------------------------------------------------------------
Invalid operation definition in line no. 74: (null)
-------------------------------------------------------------------------------
proj=latlong nadgrids=nzgd2kgrid0005.gsb ellps=GRS80                  
-------------------------------------------------------------------------------
Invalid operation definition in line no. 74: (null)
-------------------------------------------------------------------------------
proj=latlong geoidgrids=egm96_15.gtx ellps=GRS80                      
-------------------------------------------------------------------------------
Invalid operation definition in line no. 87: (null)
-------------------------------------------------------------------------------
proj=latlong geoidgrids=egm96_15.gtx ellps=GRS80                      
-------------------------------------------------------------------------------
Invalid operation definition in line no. 87: (null)
-------------------------------------------------------------------------------
proj=merc geoidgrids=egm96_15.gtx ellps=GRS80                         
-------------------------------------------------------------------------------
Invalid operation definition in line no. 96: (null)
-------------------------------------------------------------------------------
proj=merc geoidgrids=egm96_15.gtx ellps=GRS80                         
-------------------------------------------------------------------------------
Invalid operation definition in line no. 96: (null)
-------------------------------------------------------------------------------
proj=latlong geoidgrids=egm96_15.gtx axis=neu ellps=GRS80             
-------------------------------------------------------------------------------
Invalid operation definition in line no. 111: (null)
-------------------------------------------------------------------------------
proj=latlong geoidgrids=egm96_15.gtx axis=neu ellps=GRS80             
-------------------------------------------------------------------------------
Invalid operation definition in line no. 111: (null)
-------------------------------------------------------------------------------
proj=latlong geoidgrids=egm96_15.gtx axis=dne ellps=GRS80             
-------------------------------------------------------------------------------
Invalid operation definition in line no. 120: (null)
-------------------------------------------------------------------------------
proj=latlong geoidgrids=egm96_15.gtx axis=dne ellps=GRS80             
-------------------------------------------------------------------------------
Invalid operation definition in line no. 120: (null)
-------------------------------------------------------------------------------
proj=merc geoidgrids=egm96_15.gtx ellps=GRS80                         
-------------------------------------------------------------------------------
Invalid operation definition in line no. 129: (null)
-------------------------------------------------------------------------------
proj=merc geoidgrids=egm96_15.gtx ellps=GRS80                         
-------------------------------------------------------------------------------
Invalid operation definition in line no. 129: (null)
-------------------------------------------------------------------------------
proj=latlong geoidgrids=egm96_15.gtx axis=nue ellps=GRS80             
-------------------------------------------------------------------------------
Invalid operation definition in line no. 143: (null)
-------------------------------------------------------------------------------
proj=latlong geoidgrids=egm96_15.gtx axis=nue ellps=GRS80             
-------------------------------------------------------------------------------
Invalid operation definition in line no. 143: (null)
-------------------------------------------------------------------------------
proj=merc geoidgrids=egm96_15.gtx axis=sue ellps=GRS80                
-------------------------------------------------------------------------------
Invalid operation definition in line no. 152: (null)
-------------------------------------------------------------------------------
proj=merc geoidgrids=egm96_15.gtx axis=sue ellps=GRS80                
-------------------------------------------------------------------------------
Invalid operation definition in line no. 152: (null)
-------------------------------------------------------------------------------
total:  0 tests succeeded,  19 tests FAILED!
-------------------------------------------------------------------------------
Reading file 'gie/GDA.gie'
-------------------------------------------------------------------------------
total:  3 tests succeeded,   0 tests failed.
-------------------------------------------------------------------------------
Reading file 'gie/axisswap.gie'
-------------------------------------------------------------------------------
proj=pipeline step proj=latlong step proj=axisswap order=1,2,3,4 angul...
-------------------------------------------------------------------------------
Invalid operation definition in line no. 73: major axis or radius = 0 or not given
-------------------------------------------------------------------------------
proj=pipeline step proj=latlong step proj=axisswap order=-2,-1,3,4 ang...
-------------------------------------------------------------------------------
Invalid operation definition in line no. 83: major axis or radius = 0 or not given
-------------------------------------------------------------------------------
total: 25 tests succeeded,   2 tests FAILED!
-------------------------------------------------------------------------------
Reading file 'gie/builtins.gie'
-------------------------------------------------------------------------------
+proj=geocent ellps=GRS80 +lat_1=0.5 +lat_2=2                         
-------------------------------------------------------------------------------
Invalid operation definition in line no. 1233: (null)
-------------------------------------------------------------------------------
+proj=geocent ellps=GRS80 +lat_1=0.5 +lat_2=2                         
-------------------------------------------------------------------------------
Invalid operation definition in line no. 1233: (null)
-------------------------------------------------------------------------------
+proj=geocent ellps=GRS80 +lat_1=0.5 +lat_2=2                         
-------------------------------------------------------------------------------
Invalid operation definition in line no. 1233: (null)
-------------------------------------------------------------------------------
+proj=geocent ellps=GRS80 +lat_1=0.5 +lat_2=2                         
-------------------------------------------------------------------------------
Invalid operation definition in line no. 1233: (null)
-------------------------------------------------------------------------------
+proj=geocent ellps=GRS80 +lat_1=0.5 +lat_2=2                         
-------------------------------------------------------------------------------
Invalid operation definition in line no. 1233: (null)
-------------------------------------------------------------------------------
+proj=geocent ellps=GRS80 +lat_1=0.5 +lat_2=2                         
-------------------------------------------------------------------------------
Invalid operation definition in line no. 1233: (null)
-------------------------------------------------------------------------------
+proj=geocent ellps=GRS80 +lat_1=0.5 +lat_2=2                         
-------------------------------------------------------------------------------
Invalid operation definition in line no. 1233: (null)
-------------------------------------------------------------------------------
+proj=geocent ellps=GRS80 +lat_1=0.5 +lat_2=2                         
-------------------------------------------------------------------------------
Invalid operation definition in line no. 1233: (null)
-------------------------------------------------------------------------------
total: 1307 tests succeeded,   8 tests FAILED!
-------------------------------------------------------------------------------
Reading file 'gie/deformation.gie'
-------------------------------------------------------------------------------
+proj=deformation +xy_grids=alaska +z_grids=egm96_15.gtx +t_epoch=2016...
-------------------------------------------------------------------------------
Invalid operation definition in line no. 17: no system list, errno: 12

-------------------------------------------------------------------------------
+proj=deformation +xy_grids=alaska +z_grids=egm96_15.gtx +t_epoch=2016...
-------------------------------------------------------------------------------
Invalid operation definition in line no. 28: no system list, errno: 12

-------------------------------------------------------------------------------
total:  5 tests succeeded,   4 tests FAILED!
-------------------------------------------------------------------------------
Reading file 'gie/ellipsoid.gie'
-------------------------------------------------------------------------------
proj=merc                                                             
-------------------------------------------------------------------------------
Invalid operation definition in line no. 64: major axis or radius = 0 or not given
-------------------------------------------------------------------------------
total: 29 tests succeeded,   1 tests FAILED!
-------------------------------------------------------------------------------
Reading file 'gie/more_builtins.gie'
-------------------------------------------------------------------------------
proj=pipeline step init=epsg:25832 inv step init=epsg:25833 step init=...
-------------------------------------------------------------------------------
Invalid operation definition in line no. 134: no arguments in initialization list
-------------------------------------------------------------------------------
proj=pipeline step init=epsg:25832 inv step init=epsg:25833 step init=...
-------------------------------------------------------------------------------
Invalid operation definition in line no. 134: no arguments in initialization list
-------------------------------------------------------------------------------
proj=pipeline inv step proj=urm5 n=0.5 inv                            
-------------------------------------------------------------------------------
Invalid operation definition in line no. 156: major axis or radius = 0 or not given
-------------------------------------------------------------------------------
proj=pipeline step proj=urm5 n=0.5                                    
-------------------------------------------------------------------------------
Invalid operation definition in line no. 161: major axis or radius = 0 or not given
-------------------------------------------------------------------------------
proj=vgridshift grids=egm96_15.gtx ellps=GRS80                        
-------------------------------------------------------------------------------
Invalid operation definition in line no. 183: failed to load datum shift file
-------------------------------------------------------------------------------
proj=hgridshift +grids=nzgd2kgrid0005.gsb ellps=GRS80                 
-------------------------------------------------------------------------------
Invalid operation definition in line no. 214: failed to load datum shift file
-------------------------------------------------------------------------------
proj=hgridshift +grids=nzgd2kgrid0005.gsb ellps=GRS80                 
-------------------------------------------------------------------------------
Invalid operation definition in line no. 214: failed to load datum shift file
pj_cart_selftest fails with 50
-------------------------------------------------------------------------------
total: 40 tests succeeded,  10 tests FAILED!
-------------------------------------------------------------------------------
Reading file 'gie/unitconvert.gie'
-------------------------------------------------------------------------------
total:  4 tests succeeded,   0 tests failed.
-------------------------------------------------------------------------------
Grand total: 1457. Success: 1413, Failure: 44
-------------------------------------------------------------------------------
```

**Comment by kbevers** [comment]
Yeah, most of these seem to be failing these are all failing because of missing grids. I don't know if this is something we should change or not. I'll think about it during the weekend.

**Comment by cffk** [comment]
I suggest either (1) check for the grid before including the test or (2) incorporate a check for the grid and part of the test (and "succeed" if the grid is not present.  (1) is the less confusing option.

**Comment by kbevers** [comment]
@cffk Could you give an example on how you would do (1)?

In https://github.com/OSGeo/proj.4/pull/773 I have added targets for each gie test case. I have an idea that will make things simpler, but I can't figure out what the best way to test for the grids each test needs.

**Comment by sebastic** [comment]
What about defining a subset of tests that work with the grids included in the latest (1.6) proj-datumgrid release, and making all other tests opt-in with configure options and/or environment variables.

It doesn't make sense to run tests as part of the proj build that require grids that aren't included in the proj(-datumgrid). Also note that unpacking the proj-datumgrid archive into the `nad/` directory of the proj source is no longer required as it was before, it can now be unpacked into the `${PROJ_LIB}` path after proj is built & installed.

**Comment by kbevers** [comment]
> What about defining a subset of tests that work with the grids included in the latest (1.6) proj-datumgrid release, and making all other tests opt-in with configure options and/or environment variables.

Good idea. I'll see if I can figure out how to do that. Something like `./configure --with-optional-tests` perhaps?

**Comment by cffk** [comment]
Well I only know how to do it with cmake!  In the GeographicLib source distribution, the file `tools/tests.cmake user` contains constructs like:

    if (EXISTS "${_DATADIR}/geoids/egm96-5.pgm")
      # Check fix for single-cell cache bug found 2010-11-23
      add_test (NAME GeoidEval0 COMMAND GeoidEval
       -n egm96-5 --input-string "0d1 0d1;0d4 0d4")
      set_tests_properties (GeoidEval0 PROPERTIES PASS_REGULAR_EXPRESSION
        "^17\\.1[56]..\n17\\.1[45]..")
    endif ()

(I confess that I never added the tests to the autoconf setup for GeographicLib.)

**Comment by sebastic** [comment]
> Something like ./configure --with-optional-tests perhaps?

That's probably too generic.

I think you need multiple options to enabled additional tests that require grid files.

 * --with-datumgrid-tests (enable tests that require grids from proj-datumgrid in `nad/`)
 * --with-vdatumgrid-tests (enable tests that require grids from http://download.osgeo.org/proj/vdatum/)

By default none of the tests that require datum grids should be enabled because having them available in `nad/` is no longer required.

**Comment by kbevers** [comment]
> I think you need multiple options to enabled additional tests that require grid files.
> 
> --with-datumgrid-tests (enable tests that require grids from proj-datumgrid in nad/)
> --with-vdatumgrid-tests (enable tests that require grids from http://download.osgeo.org/proj/vdatum/)

That's a good point. Next question then is, how do I make that work? Is there a good example somewhere I can work of off? This is uncharted territory for me :)

**Comment by sebastic** [comment]
One option may be to change the `.gie` files to `.in` templates and have configure & CMake generate the `.gie` files based on the configure options.

Another option may be to create separate `.gie` files and use a bunch of conditionals in `test/Makefile.am` & `test/CMakeLists.txt` to only execute the tests that were selected with the configure options.

**Comment by rouault** [comment]
Wouldn't be another solution that the gie "engine" detects that a steps requires grids and check if the files are there or not ?

**Comment by kbevers** [comment]
> Wouldn't be another solution that the gie "engine" detects that a steps requires grids and check if the files are there or not ?

That's also a possibility. Although I am not sure just how. Perhaps by adding an "allowed failure" token? I'd certainly be more comfortable implementing this approach.

**Comment by busstoptaktik** [comment]
I agree with the procedure proposed by @rouault and fully designed by @kbevers in issue #774. I'll get to work on that as soon as PR #771, PR #775 and a related follow-up to #771 are done.

**Comment by kbevers** [comment]
This should be fixed as of #776.
